// @terra-money/terrain/lib/commands/contract/instantiate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const command_1 = require("@oclif/command");
const terra_js_1 = require("@terra-money/terra.js");
const config_1 = require("../../config");
const deployment_1 = require("../../lib/deployment");
const signer_1 = require("../../lib/signer");
const flag = tslib_1.__importStar(require("../../lib/flag"));
class ContractInstantiate extends command_1.Command {
    async run() {
        const { args, flags } = this.parse(ContractInstantiate);
        const connections = config_1.loadConnections(flags['config-path']);
        const config = config_1.loadConfig(flags['config-path']);
        const conf = config(flags.network, args.contract);
        // @ts-ignore
        const lcd = new terra_js_1.LCDClient(connections(flags.network));
        const signer = signer_1.getSigner({
            network: flags.network,
            signerId: flags.signer,
            keysPath: flags['keys-path'],
            lcd,
        });
        const admin = signer.key.accAddress;
        await deployment_1.instantiate({
            conf,
            signer,
            admin,
            contract: args.contract,
            codeId: flags['code-id'],
            network: flags.network,
            instanceId: flags['instance-id'],
            refsPath: flags['refs-path'],
            lcd,
        });
    }
}
exports.default = ContractInstantiate;
ContractInstantiate.description = 'Instantiate the contract.';
ContractInstantiate.flags = Object.assign({ signer: flag.signer, network: flag.network, 'instance-id': command_1.flags.string({ default: 'default' }), 'code-id': command_1.flags.integer({
        description: 'specific codeId to instantiate',
    }) }, flag.terrainPaths);
ContractInstantiate.args = [{ name: 'contract', required: true }];
