// @terra-money/terrain/lib/commands/contract/store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const command_1 = require("@oclif/command");
const terra_js_1 = require("@terra-money/terra.js");
const config_1 = require("../../config");
const deployment_1 = require("../../lib/deployment");
const signer_1 = require("../../lib/signer");
const flag = tslib_1.__importStar(require("../../lib/flag"));
class CodeStore extends command_1.Command {
    async run() {
        const { args, flags } = this.parse(CodeStore);
        const connections = config_1.loadConnections(flags['config-path']);
        const config = config_1.loadConfig(flags['config-path']);
        const conf = config(flags.network, args.contract);
        // @ts-ignore
        const lcd = new terra_js_1.LCDClient(connections(flags.network));
        const signer = signer_1.getSigner({
            network: flags.network,
            signerId: flags.signer,
            keysPath: flags['keys-path'],
            lcd,
        });
        await deployment_1.storeCode({
            conf,
            noRebuild: flags['no-rebuild'],
            contract: args.contract,
            signer,
            network: flags.network,
            refsPath: flags['refs-path'],
            lcd,
            codeId: flags['code-id'],
        });
    }
}
exports.default = CodeStore;
CodeStore.description = 'Store code on chain.';
CodeStore.flags = Object.assign({ signer: flag.signer, network: flag.network, 'no-rebuild': flag.noRebuild, 'code-id': command_1.flags.integer({}) }, flag.terrainPaths);
CodeStore.args = [{ name: 'contract', required: true }];
