// @terra-money/terrain/lib/lib/flag.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.terrainPaths = exports.frontendRefsPath = exports.keysPath = exports.refsPath = exports.configPath = exports.network = exports.instanceId = exports.noRebuild = exports.signer = void 0;
const command_1 = require("@oclif/command");
exports.signer = command_1.flags.string({ default: 'test1' });
exports.noRebuild = command_1.flags.boolean({
    description: 'deploy the wasm bytecode as is.',
    default: false,
});
exports.instanceId = command_1.flags.string({ default: 'default', description: 'enable management of multiple instances of the same contract' });
exports.network = command_1.flags.string({ default: 'localterra', description: 'network to deploy to from config.terrain.json' });
exports.configPath = command_1.flags.string({ default: './config.terrain.json' });
exports.refsPath = command_1.flags.string({ default: './refs.terrain.json' });
exports.keysPath = command_1.flags.string({ default: './keys.terrain.js' });
exports.frontendRefsPath = command_1.flags.string({
    default: './frontend/src/',
});
// These three are commonly used together.
exports.terrainPaths = {
    'config-path': exports.configPath,
    'refs-path': exports.refsPath,
    'keys-path': exports.keysPath,
};
